### PR TITLE
Fix for memory leaks

### PR DESCRIPTION
**Pull request by NomAnor**
Valgrind's memcheck showed me some memory leaks when running the wren cli so I tried to fix them.

The first is in the cli itself. It's not a serious one because the memory gets freed a soon a the programm exits. The memory allocated for the root directory path is not freed.

In the wren compiler the constants buffer data's ownership is not transfered to the new function but the constans are copied so the buffer's data must be freed.

The last one is debatable. According to the C Standard `realloc(NULL, size)` is the same as `malloc(size)`. When the memory for the internal data structures is allocated the previous memory (which defaults to NULL) is deallocated by calling `DEALLOCATE(vm, NULL)` this will invoke `realloc(NULL, 0)` which will behave like `malloc(0)` which can return (implementation defined) a pointer to a zero sized object which must not be dereferenced, but freed. By using the DEALLOCATE macro with a NULL ptr these objects leak (they are size 0 so not really a memory leak). I changed the default reallocate function to one which does the right thing.

### CODE REVIEW

**Comment by BarabasGitHub** [comment]
I think it's safe to do free(NULL). So you probably don't have to check for that.

**Comment by NomAnor** [comment]
It is safe, but the checks are necessary because `realloc(NULL, 0)` is not equivalent to `free(NULL)`.

There are 4 possible invocations for the wren reallocation function:

  * `ptr == NULL && size == 0`: Do nothing because the function is asked to deallocate a NULL pointer, thats not the behaviour of realloc (it's equivalent to `malloc(0)`, which is implementation defined)
  * `ptr != NULL && size == 0`: Deallocate memory. It's implementation defined wether realloc only frees the memory or reallocates a new zero sized object.
  * `ptr == NULL && size > 0`: Allocate memory, handled by realloc
  * `ptr != NULL && size > 0`: Reallocate memory, handled by realloc

See http://en.cppreference.com/w/c/memory/realloc, same for malloc

> If new_size is zero, the behavior is implementation defined (null pointer may be returned (in which case the old memory block is not freed), or some non-null pointer may be returned that may not be used to access storage).

**Comment by BarabasGitHub** [comment]
Ah, I see. So whenever the size == 0 free has to be called instead of
realloc. I understand. =)

File: src/cli/io.c [hunk 1]
```diff
@@ -46,6 +46,10 @@ char* readFile(const char* path)
 
 void setRootDirectory(const char* path)
 {
+  if (rootDirectory != NULL)
+  {
+    free((void*)rootDirectory);
+  }
```
Comment (bjorn) [inline src/cli/io.c]: Regardless, here an unnecessary NULL check is done before calling free.

**Comment by munificent** [comment]
This is awesome! I really appreciate you doing this. I've tried to get Valgrind running on my machine before but was never successful.

I made a few changes to your patch but otherwise merged it in. Thanks!

If you want, send me a PR to add yourself to the AUTHORS file. :grinning:

**Pull request merged**